File: packages/python-sdk/src/chromiumfish/fetch.py
"""Download, verify, and cache the ChromiumFish browser build.

Fetch model: resolve ``version × platform`` to a GitHub
Release asset, verify its SHA-256, extract it to a per-version cache dir, and
return the path to the launchable binary.
"""
from __future__ import annotations

import hashlib
import os
import platform
import shutil
import subprocess
import sys
import tarfile
import urllib.request
import zipfile
from pathlib import Path

from .version import browser_version, release_base_url


class UnsupportedPlatformError(RuntimeError):
    pass


def cache_root() -> Path:
    env = os.environ.get("CHROMIUMFISH_CACHE_DIR")
    if env:
        return Path(env).expanduser()
    if sys.platform == "darwin":
        return Path.home() / "Library" / "Caches" / "chromiumfish"
    if os.name == "nt":
        base = os.environ.get("LOCALAPPDATA", str(Path.home() / "AppData" / "Local"))
        return Path(base) / "chromiumfish"
    return Path(os.environ.get("XDG_CACHE_HOME", str(Path.home() / ".cache"))) / "chromiumfish"


def platform_slug() -> str:
    """e.g. ``linux-x64``, ``win-x64``, ``mac-arm64``."""
    machine = platform.machine().lower()
    arch = {
        "x86_64": "x64", "amd64": "x64",
        "aarch64": "arm64", "arm64": "arm64",
    }.get(machine)
    if arch is None:
        raise UnsupportedPlatformError(f"unsupported architecture: {machine}")
    if sys.platform.startswith("linux"):
        return f"linux-{arch}"
    if sys.platform == "darwin":
        return f"mac-{arch}"
    if os.name == "nt":
        return f"win-{arch}"
    raise UnsupportedPlatformError(f"unsupported platform: {sys.platform}")


def _asset_name(version: str) -> str:
    slug = platform_slug()
    ext = "zip" if slug.startswith("win") else "tar.gz"
    return f"chromiumfish-{version}-{slug}.{ext}"


def _binary_name() -> str:
    if os.name == "nt":
        return "chromiumfish.exe"
    return "chromiumfish"  # falls back to "chrome" during discovery


def install_dir(version: str | None = None) -> Path:
    version = version or browser_version()
    return cache_root() / version / platform_slug()


def find_binary(root: Path) -> Path | None:
    """Locate the launchable binary inside an extracted build."""
    candidates = ["chromiumfish", "chrome", "chromiumfish.exe", "chrome.exe", "ChromiumFish"]
    for name in candidates:
        direct = root / name
        if direct.is_file():
            return direct
    for name in candidates:
        for hit in root.rglob(name):
            if hit.is_file():
                return hit
    return None


def _download(url: str, dest: Path) -> None:
    dest.parent.mkdir(parents=True, exist_ok=True)
    print(f"[chromiumfish] downloading {url}", file=sys.stderr)
    with urllib.request.urlopen(url) as resp, open(dest, "wb") as out:  # noqa: S310
        total = int(resp.headers.get("Content-Length", 0))
        read = 0
        while chunk := resp.read(1 << 20):
            out.write(chunk)
            read += len(chunk)
            if total:
                pct = read * 100 // total
                print(f"\r[chromiumfish] {pct:3d}%  ({read >> 20} / {total >> 20} MiB)",
                      end="", file=sys.stderr)
        print("", file=sys.stderr)


def _sha256(path: Path) -> str:
    h = hashlib.sha256()
    with open(path, "rb") as f:
        while chunk := f.read(1 << 20):
            h.update(chunk)
    return h.hexdigest()


def _verify(archive: Path, base_url: str, asset: str) -> None:
    try:
        with urllib.request.urlopen(f"{base_url}/{asset}.sha256") as r:  # noqa: S310
            expected = r.read().decode().split()[0].strip()
    except Exception:  # noqa: BLE001
        print("[chromiumfish] warning: no .sha256 published, skipping verification", file=sys.stderr)
        return
    actual = _sha256(archive)
    if actual != expected:
        archive.unlink(missing_ok=True)
        raise RuntimeError(f"checksum mismatch for {asset}: {actual} != {expected}")


def _extract(archive: Path, dest: Path) -> None:
    dest.mkdir(parents=True, exist_ok=True)
    if archive.name.endswith(".zip"):
        with zipfile.ZipFile(archive) as z:
            z.extractall(dest)
    else:
        with tarfile.open(archive) as t:
            t.extractall(dest)  # noqa: S202 - trusted first-party asset


def _macos_prepare(target: Path) -> None:
    """Identity-clean macOS prep.

    1. Strip the ``com.apple.quarantine`` flag. Programmatic downloads usually
       don't set it, but browsers/tools might — removing it avoids Gatekeeper's
       "unidentified developer" block without any notarization.
    2. Ensure the bundle is ad-hoc signed (``codesign -s -``). Apple Silicon
       refuses to run an unsigned binary; ad-hoc signing fixes that and embeds
       NO certificate, name, or identity. (Release builds ship ad-hoc signed;
       this is a defensive fallback.)
    """
    if sys.platform != "darwin":
        return
    app = next(target.glob("*.app"), None)
    sign_target = app or find_binary(target)
    subprocess.run(["xattr", "-dr", "com.apple.quarantine", str(target)],
                   check=False, capture_output=True)
    if sign_target:
        valid = subprocess.run(["codesign", "--verify", "--quiet", str(sign_target)],
                               capture_output=True).returncode == 0
        if not valid:
            subprocess.run(["codesign", "--force", "--deep", "--sign", "-", str(sign_target)],
                           check=False, capture_output=True)


def fetch(version: str | None = None, *, force: bool = False) -> Path:
    """Ensure the browser build is present and return the binary path."""
    version = version or browser_version()
    target = install_dir(version)

    if force and target.exists():
        shutil.rmtree(target, ignore_errors=True)

    if target.exists():
        binp = find_binary(target)
        if binp:
            return binp

    base = release_base_url(version)
    asset = _asset_name(version)
    archive = cache_root() / version / asset
    _download(f"{base}/{asset}", archive)
    _verify(archive, base, asset)
    _extract(archive, target)
    archive.unlink(missing_ok=True)
    _macos_prepare(target)

    binp = find_binary(target)
    if not binp:
        raise RuntimeError(f"no browser binary found in extracted build at {target}")
    if os.name != "nt":
        binp.chmod(0o755)
    print(f"[chromiumfish] ready: {binp}", file=sys.stderr)
    return binp


def binary_path(version: str | None = None, *, download: bool = True) -> Path:
    """Path to the cached binary, fetching it if needed (and allowed)."""
    version = version or browser_version()
    existing = find_binary(install_dir(version))
    if existing:
        return existing
    if not download:
        raise FileNotFoundError(
            f"ChromiumFish {version} not installed. Run `chromiumfish fetch`."
        )
    return fetch(version)

File: packages/js-sdk/src/fetch.ts
/**
 * Download, verify, and cache the ChromiumFish browser build.
 *
 * Resolves `version × platform` to a GitHub Release asset, verifies its
 * SHA-256, extracts it to a per-version cache dir, and returns the path to the
 * launchable binary.
 */
import { createHash } from "node:crypto";
import { spawnSync } from "node:child_process";
import * as fs from "node:fs";
import * as https from "node:https";
import * as os from "node:os";
import * as path from "node:path";
import { assertSafeVersion, browserVersion, releaseBaseUrl } from "./version.js";

export function cacheRoot(): string {
  const env = process.env.CHROMIUMFISH_CACHE_DIR;
  if (env) return env;
  if (process.platform === "darwin")
    return path.join(os.homedir(), "Library", "Caches", "chromiumfish");
  if (process.platform === "win32")
    return path.join(process.env.LOCALAPPDATA || path.join(os.homedir(), "AppData", "Local"), "chromiumfish");
  return path.join(process.env.XDG_CACHE_HOME || path.join(os.homedir(), ".cache"), "chromiumfish");
}

export function platformSlug(): string {
  const arch = ({ x64: "x64", arm64: "arm64" } as Record<string, string>)[process.arch];
  if (!arch) throw new Error(`unsupported architecture: ${process.arch}`);
  if (process.platform === "linux") return `linux-${arch}`;
  if (process.platform === "darwin") return `mac-${arch}`;
  if (process.platform === "win32") return `win-${arch}`;
  throw new Error(`unsupported platform: ${process.platform}`);
}

function assetName(version: string): string {
  assertSafeVersion(version);
  const slug = platformSlug();
  const ext = slug.startsWith("win") ? "zip" : "tar.gz";
  return `chromiumfish-${version}-${slug}.${ext}`;
}

export function installDir(version = browserVersion()): string {
  return path.join(cacheRoot(), assertSafeVersion(version), platformSlug());
}

const BINARY_NAMES = ["chromiumfish", "chrome", "chromiumfish.exe", "chrome.exe", "ChromiumFish"];

export function findBinary(root: string): string | null {
  if (!fs.existsSync(root)) return null;
  for (const name of BINARY_NAMES) {
    const direct = path.join(root, name);
    // statSync can throw if the file is removed between existsSync and stat
    // (TOCTOU); treat any stat failure as "not a usable binary here".
    try {
      if (fs.statSync(direct).isFile()) return direct;
    } catch {
      /* keep looking */
    }
  }
  const stack = [root];
  while (stack.length) {
    const dir = stack.pop()!;
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) stack.push(full);
      else if (BINARY_NAMES.includes(entry.name)) return full;
    }
  }
  return null;
}

// Idle-timeout (ms) applied to every download/fetch socket. A stalled server
// (no bytes for this long) aborts instead of hanging the launch forever.
const DOWNLOAD_IDLE_TIMEOUT_MS = 60_000;
const FETCH_IDLE_TIMEOUT_MS = 30_000;

function download(url: string, dest: string): Promise<void> {
  fs.mkdirSync(path.dirname(dest), { recursive: true });
  return new Promise((resolve, reject) => {
    const get = (u: string) => {
      const req = https.get(u, (res) => {
        if (res.statusCode && res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
          res.resume();
          return get(res.headers.location);
        }
        if (res.statusCode !== 200) {
          res.resume();
          return reject(new Error(`download failed (${res.statusCode}) for ${u}`));
        }
        const total = Number(res.headers["content-length"] || 0);
        let read = 0;
        const out = fs.createWriteStream(dest);
        // On any failure: tear down both streams and remove the partial file
        // so a later run doesn't trip over a truncated/corrupt download.
        const fail = (err: Error) => {
          res.destroy();
          out.destroy();
          try { fs.rmSync(dest, { force: true }); } catch { /* best effort */ }
          reject(err);
        };
        res.on("data", (c) => {
          read += c.length;
          if (total) process.stderr.write(`\r[chromiumfish] ${Math.floor((read / total) * 100)}%`);
        });
        res.on("error", fail);
        res.pipe(out);
        out.on("finish", () => { process.stderr.write("\n"); out.close(() => resolve()); });
        out.on("error", fail);
      });
      req.on("error", reject);
      req.setTimeout(DOWNLOAD_IDLE_TIMEOUT_MS, () => {
        req.destroy(new Error(`download timed out (no data for ${DOWNLOAD_IDLE_TIMEOUT_MS}ms) for ${u}`));
      });
    };
    process.stderr.write(`[chromiumfish] downloading ${url}\n`);
    get(url);
  });
}

function fetchText(url: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const get = (u: string) => {
      const req = https.get(u, (res) => {
        if (res.statusCode && res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
          res.resume();
          return get(res.headers.location);
        }
        if (res.statusCode !== 200) { res.resume(); return reject(new Error(`HTTP ${res.statusCode}`)); }
        let data = "";
        res.setEncoding("utf8");
        res.on("data", (c) => (data += c));
        res.on("end", () => resolve(data));
        res.on("error", reject);
      });
      req.on("error", reject);
      req.setTimeout(FETCH_IDLE_TIMEOUT_MS, () => {
        req.destroy(new Error(`request timed out for ${u}`));
      });
    };
    get(url);
  });
}

function sha256(file: string): string {
  return createHash("sha256").update(fs.readFileSync(file)).digest("hex");
}

function extract(archive: string, dest: string): void {
  fs.mkdirSync(dest, { recursive: true });
  // Modern tar (incl. Windows 10+ bsdtar) extracts both .tar.gz and .zip.
  const r = spawnSync("tar", ["-xf", archive, "-C", dest], { stdio: "inherit" });
  if (r.status !== 0) throw new Error(`extraction failed for ${archive}`);
}

export async function fetchBrowser(version = browserVersion(), force = false): Promise<string> {
  const target = installDir(version);
  if (force && fs.existsSync(target)) fs.rmSync(target, { recursive: true, force: true });

  const cached = findBinary(target);
  if (cached) return cached;

  const base = releaseBaseUrl(version);
  const asset = assetName(version);
  const archive = path.join(cacheRoot(), version, asset);

  await download(`${base}/${asset}`, archive);

  try {
    const expected = (await fetchText(`${base}/${asset}.sha256`)).trim().split(/\s+/)[0];
    const actual = sha256(archive);
    if (actual !== expected) {
      fs.rmSync(archive, { force: true });
      throw new Error(`checksum mismatch for ${asset}: ${actual} !== ${expected}`);
    }
  } catch (e: any) {
    if (String(e?.message || e).includes("HTTP"))
      process.stderr.write("[chromiumfish] warning: no .sha256 published, skipping verification\n");
    else throw e;
  }

  extract(archive, target);
  fs.rmSync(archive, { force: true });
  macosPrepare(target);

  const bin = findBinary(target);
  if (!bin) throw new Error(`no browser binary found in extracted build at ${target}`);
  if (process.platform !== "win32") fs.chmodSync(bin, 0o755);
  process.stderr.write(`[chromiumfish] ready: ${bin}\n`);
  return bin;
}

/**
 * Identity-clean macOS prep: strip the download
 * quarantine flag, and ensure an ad-hoc signature so Apple Silicon will run the
 * binary — no certificate, name, or identity attached. Release builds ship
 * ad-hoc signed; this is a defensive fallback.
 */
function macosPrepare(target: string): void {
  if (process.platform !== "darwin") return;
  spawnSync("xattr", ["-dr", "com.apple.quarantine", target], { stdio: "ignore" });
  const app = fs.readdirSync(target).find((n) => n.endsWith(".app"));
  const signTarget = app ? path.join(target, app) : findBinary(target);
  if (signTarget) {
    const valid = spawnSync("codesign", ["--verify", "--quiet", signTarget]).status === 0;
    if (!valid) spawnSync("codesign", ["--force", "--deep", "--sign", "-", signTarget], { stdio: "ignore" });
  }
}

export async function binaryPath(version = browserVersion(), download = true): Promise<string> {
  const existing = findBinary(installDir(version));
  if (existing) return existing;
  if (!download) throw new Error(`ChromiumFish ${version} not installed. Run \`npx chromiumfish fetch\`.`);
  return fetchBrowser(version);
}
